text/0000-command-ergonomics.md

```diff
@@ -0,0 +1,542 @@
+
+We will introduce new APIs on `Command`,
+for running the command and collecting its output:
+
+```rust
+impl Command {
+    fn run(&mut self) -> Result<(), SubprocessError>;
+    fn get_output_bytes(&mut self) -> Result<Vec<u8>, SubprocessError>;
+    fn get_output(&mut self) -> Result<String, SubprocessError>;
+    fn get_output_line(&mut self) -> Result<String, SubprocessError>;
+    fn get_output_read(&mut self) -> impl std::io::Read;
+}
+struct SubprocessError { ... }
+impl From<SubprocessError> for io::Error { ... }
+```
+
+The `.output()` function and `std::process::Output`
+will be deprecated.
+
+No significant changes are made to the `Command` construction APIs,
+but it may become necessary to call `.stderr()` explicitly
+to use the new methods (see Unresolved Questions).
+
+## Use cases
+
+We aim to serve well each of the following people:
+
+ * Alice writes a CLI utility to orchestrate processes and wants top-notch error reporting from subprocesses.
+
+ * Bob migrates ad-hoc automation scripts from bash to Rust.
+
+ * Carol writes a generic application and just wants all errors to be maximally useful by default.
+
+ * Dionysus wants to run `diff`, which exits `0` for "no difference",
+   `1` for "difference found" and
+   another value for failure.
+
+(Partly cribbed from one of
+@matklad's [comments](https://github.com/rust-lang/rust/pull/89004#issuecomment-923803209) in #89004)
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## New methods on `Command`:
+
+ * `fn run(&mut self) -> Result<(), SubprocessError>`:
+
+   Runs the command.
+   Equivalent to `.spawn()` followed by `.status()`,
+   but with better error handling.
+
+ * `fn get_output_bytes(&mut self) -> Result<Vec<u8>, SubprocessError>`:
+
+   Runs the command and collects its stdout.
+   After the child indicates EOF on its stdout,
+   we will wait for it to finish and check the exit status.
+
+ * `fn get_output(&mut self) -> Result<Vec<String>, SubprocessError>`:
+
+   Runs the command and collects its stdout.
+   Decodes the stdout as UTF-8, and fails if that's not possible.
+   Does not trim any trailing line ending.
+
+ * `fn get_output_line(&mut self) -> Result<Vec<String>, SubprocessError>`:
+
+   Runs the command and collects its stdout.
+   Decodes the stdout as UTF-8, and fails if that's not possible.
+   Fails unless the output is a single line (with or without line ending).
+   Trims the line ending (if any).
+
+ * `fn get_output_read(&mut self) -> std::process::ChildOutputStream`
+   (where `struct ChildOutputStream` implements `io::Read`
+   and is `Send + Sync + 'static`).
+
+   Starts the command, allowing the caller to
+   read the stdout in a streaming way.
+   Neither EOF nor an error will be reported by `ChildOutputStream`
+   until the child has exited, *and* the stdout pipe reports EOF.
+   (This includes errors due to nonempty stderr,
+   if stderr was set to `piped`.)
```

Comment:
This description makes me wonder about the semantics of the *other* functions in this family. Do they wait for process exit, or for process exit *and* EOF? (I'm hoping the former, or possibly "process exit and then a non-blocking read saying it would block".)

Reply:
This is stated explicitly in the docs for what I am now calling `read_stdout_bytes`.  All of these functions wait for both EOF and process exit.  I will add a short xref note to the other two functions.  In the real docs the spec text should probably be cut and pasted instead.